### PR TITLE
refactor: refactor shortcode parsing logic for improved reliability

### PR DESCRIPTION
**Pull request by aoliverwd**
## Executive Summary
- Refactored the `callShortcode` method in the `Parser` class to improve argument parsing and validation.
- Enhanced support for shortcode attributes with better handling of key-value pairs.
- Improved maintainability and readability of the `callShortcode` logic.

---

## New Features
- **Enhanced Shortcode Parsing**: The `callShortcode` method now supports more robust parsing of shortcode attributes, enabling better handling of key-value pairs in the syntax.

---

## Improvements
- **Refactored Argument Parsing**:
  - Replaced manual string manipulation with regex-based parsing for shortcodes.
  - Improved validation of shortcode arguments to ensure correctness before processing.
- **Attribute Handling**:
  - Added support for extracting and validating attributes from shortcode syntax using regex.
  - Simplified the merging of attributes with the global dataset for better maintainability.

---

## Bug Fixes
- **Shortcode Argument Validation**: Fixed issues where invalid or incomplete shortcode arguments could lead to unexpected behavior or errors.

---

## Technical Changes
- **Regex-Based Parsing**:
  - Introduced regex to parse shortcode syntax and extract attributes.
  - Removed older, less efficient string manipulation logic.
- **Code Simplification**:
  - Reduced redundant code for argument formatting and validation.
  - Improved readability by consolidating attribute handling logic.

---

## Breaking Changes
- **No breaking changes identified.**

---

## Risk Assessment
**Risk Level**: Low
- **Deployment Risk**: Minimal, as changes are isolated to the `callShortcode` method.
- **Rollback Complexity**: Low, as the refactor does not introduce new dependencies or alter external behavior.
- **Monitoring**: Verify that shortcodes with attributes are parsed and executed correctly.

---

## Testing & Validation
- **Test Coverage**:
  - Ensure existing tests for `callShortcode` cover the new attribute handling logic.
  - Add tests for edge cases, such as invalid or incomplete shortcode syntax.
- **QA Focus**:
  - Validate that shortcodes with attributes are processed correctly.
  - Test shortcodes with and without attributes to ensure backward compatibility.

---

## Reasoning Behind Changes
- **Improved Maintainability**: The refactor simplifies the `callShortcode` logic, making it easier to understand and maintain.
- **Enhanced Functionality**: The new attribute handling logic allows for more flexible and robust shortcode parsing.
- **Bug Prevention**: The improved validation reduces the likelihood of errors caused by invalid shortcode syntax.

---

## Notable Files Changed
- **`src/Parser.php`**:
  - Refactored the `callShortcode` method to use regex for parsing and validating shortcode attributes.
  - Simplified the merging of attributes with the global dataset.

### CODE REVIEW

File: src/Parser.php [hunk 1]
```diff
@@ -155,14 +155,20 @@ public function regShortcode(string $name, string|callable $theMethod): Parser
      */
     private function callShortcode(string $shortcodeSyntax, array $dataset): string
     {
-        $sanatise_1 = str_replace('&quot;', '"', $shortcodeSyntax);
-        $sanatise_2 = str_replace(['[', ']'], '', $sanatise_1);
-        $args = explode(' ', $sanatise_2);
+        preg_match('/^\[(.*?)\s+(.*?)\s*\]$|^\[(.*?)\]$/', $shortcodeSyntax, $args);
+        $args = array_values(array_filter($args));
```
[the remark this answers is not in the document]
Comment (aoliverwd) [reply src/Parser.php:159]: In this use case, using "0" as a method is considered bad practice and should be avoided. It should be removed.

File: src/Parser.php [hunk 2]
```diff
@@ -176,27 +182,23 @@ private function callShortcode(string $shortcodeSyntax, array $dataset): string
                 return $method . ' not found';
```
[the remark this answers is not in the document]
Comment (aoliverwd) [reply src/Parser.php]: Fixed in f14abbbcdceb1cddaf0cd4a7a719e5c272b1f4a8

File: src/Parser.php [hunk 3]
```diff
@@ -176,27 +182,23 @@ private function callShortcode(string $shortcodeSyntax, array $dataset): string
...
-                $newArg = explode('=', str_replace('"', '', $thisArg));
-
-                if (count($newArg) === 2) {
-                    $theArgs[trim($newArg[0])] = trim($newArg[1]);
+            /** Check if attributes are valid */
+            if ($attributes) {
+                foreach ($attributes as $key => $attribute) {
+                    $attributes[trim($attribute[1])] = trim($attribute[2]);
+                    unset($attributes[$key]);
                 }
```
[the remark this answers is not in the document]
Comment (aoliverwd) [reply src/Parser.php:190]: This is expected behaviour

File: src/Parser.php [hunk 4]
```diff
@@ -155,14 +155,20 @@ public function regShortcode(string $name, string|callable $theMethod): Parser
...
 
         /** Get method name from args */
-        $methodName = isset($args[0]) && isset($this->shortcode_methods[$args[0]])
-            ? $this->shortcode_methods[$args[0]]
-            : false;
+        $methodName = isset($this->shortcode_methods[$args[1]]) ? $this->shortcode_methods[$args[1]] : false;
+
+        // Get attributes from args
+        $args[2] = isset($args[2]) ? $args[2] : '';
+        preg_match_all('/(.*?)="(.*?)"/', $args[2], $attributes, PREG_SET_ORDER);
```
[the remark this answers is not in the document]
Comment (aoliverwd) [reply src/Parser.php:171]: The previous implementation also stripped `"`
